src/builder_state.rs

```diff
@@ -490,17 +492,50 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
 
             async_sleep(sleep_interval).await
         }
+
+        // Don't build an empty block
+        if self.tx_queue.is_empty() {
+            return None;
+        }
+
+        let max_block_size = self.global_state.read_arc().await.max_block_size;
+        let transactions_to_include = self.tx_queue.iter().scan(0, |total_size, tx| {
+            if *total_size >= max_block_size {
+                None
+            } else {
+                // This way we will include one transaction over our target block length.
+                // This is done on purpose, otherwise we'd have a possible failure state
+                // where a single transaction larger than target block state is stuck in
+                // queue and we just build empty blocks
+                *total_size += tx.len;
+                Some(tx.tx.clone())
+            }
+        });
+
         if let Ok((payload, metadata)) =
             <TYPES::BlockPayload as BlockPayload<TYPES>>::from_transactions(
-                self.tx_queue.iter().map(|tx| tx.tx.clone()),
+                transactions_to_include,
                 &self.validated_state,
                 &self.instance_state,
             )
             .await
         {
             let builder_hash = payload.builder_commitment(&metadata);
             // count the number of txns
-            let txn_count = self.tx_queue.len();
+            let actual_txn_count = payload.num_transactions(&metadata);
+
+            // Payload is empty despite us checking that tx_queue isn't empty earlier.
+            // This indicates that the block was truncated due to sequencer block length
+            // limits and we have a transaction too big to ever be included in the head of
+            // our queue. We need to drop it and mark as "included" so that if we receive
+            // it again we don't even bother with it.
```

Comment:
Is the "sequencer block length limits" the same as `max_block_size`? The comment and code here make sense to me, but see my first question above--I'm wondering whether we can skip the large transaction there so that we don't need the list updates here.

Reply:
Sequencer block length limits are different from our `max_block_size`. `max_block_size` is our estimation of what we can calculate VID for in time, while sequencer limits are determined by chainconfig.

Unfortunately, we can't check if transaction is over the sequencer's limit immediately, because there's no way for us to access the chainconfig, as we don't know anything about the sequencer types. Moreover, even if we did, the sequencer's maximum block size != maximum transaction size, because blocks have overhead on namespaces and other stuff. Moreover moreover, even if we had a way to get a clearly-defined transaction size from the sequencer, we can't even know the actual length of `TYPES::Transaction`!

So this code block detects if a single transaction is big enough to single-handedly go over the sequencer's block size limit by observing the following scenario:
- we have some transactions in queue
- we build a block from those by calling `<TYPES::BlockPayload as BlockPayload<TYPES>>::from_transactions`
- the definition of that for sequencer silently truncates the block on the first transaction it encounters that makes the block go over sequencer's limit
- `<TYPES::BlockPayload as BlockPayload<TYPES>>::from_transactions` returns a block, but calling `num_transactions` on it shows that it's completely empty
This means that the first transaction that we tried to include is too big, so we'll drop it from our queue.

Reply:
Thanks for the detailed explanation. Makes sense now!